Source/DafnyCore/Resolver/PreType/PreTypeResolve.State.cs

```diff
@@ -0,0 +1,886 @@
+      if (printableContext != null) {
+        PrintTypeInferenceState("(partial) " + printableContext);
+      }
+      bool anythingChanged;
+      do {
+        anythingChanged = false;
+        anythingChanged |= ApplySubtypeConstraints();
+        anythingChanged |= ApplyEqualityConstraints();
+        anythingChanged |= ApplyGuardedConstraints();
+        if (makeDecisions) {
+          if (DecideHeadsFromBounds(true)) {
+            anythingChanged = true;
+          } else if (DecideHeadsFromBounds(false)) {
+            anythingChanged = true;
+          }
+        }
+      } while (anythingChanged);
+    }
+
+    void SolveAllTypeConstraints(string printableContext) {
+      PrintTypeInferenceState(printableContext);
+      PartiallySolveTypeConstraints(null);
+
+      PartiallySolveTypeConstraints(null, true);
+
+      if (ApplyDefaultAdvice()) {
+        PartiallySolveTypeConstraints(null, true);
+      }
+
+      PrintLegend();
+      ConfirmTypeConstraints();
+      ClearState();
+    }
+
+    void ClearState() {
+      unnormalizedSubtypeConstraints.Clear();
+      guardedConstraints.Clear();
+      defaultAdvice.Clear();
+      confirmations.Clear();
+      allPreTypeProxies.Clear();
+    }
+
+    public void PrintTypeInferenceState(string/*?*/ header = null) {
+      if (!resolver.Options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      Console.WriteLine("*** Type inference state ***{0}", header == null ? "" : $" {header} ");
+      PrintList("Subtype constraints", unnormalizedSubtypeConstraints, stc => {
+        return $"{stc.Super} :> {stc.Sub}";
+      });
+      PrintList("Equality constraints", equalityConstraints, eqc => {
+        return $"{eqc.A} == {eqc.B}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_GUARDED_CONSTRAINTS
+      PrintList("Guarded constraints", guardedConstraints, gc => {
+        return gc.Kind + Util.Comma("", gc.Arguments, arg => $" {arg}");
+      });
+#endif
+      PrintList("Default-type advice", defaultAdvice, advice => {
+        return $"{advice.PreType} ~-~-> {advice.WhatString}";
+      });
+#if IS_THERE_A_GOOD_WAY_TO_PRINT_CONFIRMATIONS
+      PrintList("Post-inference confirmations", confirmations, c => {
+        return $"{TokToShortLocation(c.tok)}: {c.Check} {c.PreType}: {c.ErrorMessage()}";
+      });
+#endif
+    }
+
+    void PrintLegend() {
+      PrintList("Legend", allPreTypeProxies, pair => {
+        var s = Pad($"?{pair.Item1.UniqueId}", 4) + pair.Item1;
+        return pair.Item2 == null ? s : $"{Pad(s, 20)}  {pair.Item2}";
+      });
+    }
+
+    public static string TokToShortLocation(IToken tok) {
+      return $"{System.IO.Path.GetFileName(tok.filename)}({tok.line},{tok.col - 1})";
+    }
+
+    void PrintList<T>(string rubric, List<T> list, Func<T, string> formatter) {
+      if (!resolver.Options.Get(CommonOptionBag.NewTypeInferenceDebug)) {
+        return;
+      }
+      Console.WriteLine($"    {rubric}:");
```

Comment:
Please don't use `Console.` anywhere. You can use `resolver.Options.OutputWriter` or `resolver.Options.ErrorWriter` instead.

I would enjoy adding a tiny Roslyn plugin to our codebase to disallow using `Console.` unless you add a pragma, but haven't gotten to that.